azuredevops.go

```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
+	i importBody, project azureDevOpsProject) (wharfapi.Project, error) {
+	buildDefinitionStr, err := getBuildDefinitionWritesError(c, i, project.Name)
 	if err != nil {
-		fmt.Println("Unable to get: ", err)
-		return []byte{}, err
+		return wharfapi.Project{}, err
 	}
 
-	req.SetBasicAuth(user, token)
-
-	resp, err := http.DefaultClient.Do(req)
+	gitURL, err := getGitURL(provider, i.Group, project)
 	if err != nil {
-		fmt.Println("Unable to do request: ", err)
-		return []byte{}, err
+		fmt.Println("Unable to construct git url ", err)
+		ginutil.WriteComposingProviderDataError(c, err,
+			fmt.Sprintf("Unable to construct git url for project '%s' in group '%s'", project.Name, i.Group))
+		return wharfapi.Project{}, err
 	}
 
-	defer resp.Body.Close()
-
-	if resp.StatusCode != http.StatusOK {
-		fmt.Println("Unable to get. Status code: ", resp.StatusCode)
-		return []byte{}, err
-	}
+	projectInDb, err := client.PutProject(
+		wharfapi.Project{
+			Name:            project.Name,
+			TokenID:         i.TokenID,
+			GroupName:       i.Group,
+			BuildDefinition: buildDefinitionStr,
+			Description:     project.Description,
+			ProviderID:      provider.ProviderID,
+			GitURL:          gitURL})
 
-	bodyBytes, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		fmt.Println(err)
-		return []byte{}, err
+		fmt.Println("Unable to put project: ", err)
+		ginutil.WriteAPIClientWriteError(c, err,
+			fmt.Sprintf("Unable to import project '%s' from group '%s' at url '%s'.",
+				i.Project, i.Group, gitURL))
+		return wharfapi.Project{}, err
 	}
 
-	return bodyBytes, nil
+	return projectInDb, nil
 }
 
-func getProjectBranches(i importBody, project string) ([]azureDevOpsBranch, error) {
-	urlPath, err := url.Parse(i.URL)
+func postBranchesWritesError(c *gin.Context, client wharfapi.Client, i importBody,
+	project azureDevOpsProject, projectInDb wharfapi.Project) error {
+	repositories, err := getRepositoriesWritesError(c, i, project)
 	if err != nil {
-		fmt.Println("Unable to get url: ", err)
-		return []azureDevOpsBranch{}, err
+		fmt.Println("Unable to get repos: ", err)
+		return err
 	}
 
-	urlPath.Path = fmt.Sprintf("%v/%v/%v/%v/%v", i.Group, project, apiRepositories, project, refsPath)
-	q := url.Values{}
-	q.Add("api-version", "5.0")
-	q.Add("filter", "heads/")
-	urlPath.RawQuery = q.Encode()
+	projectBranches, err := getProjectBranchesWritesError(c, i, project.Name)
+	if err != nil {
+		fmt.Println("Unable to get project branches: ", err)
+		return err
+	}
 
-	fmt.Println(urlPath.String())
+	for _, branch := range projectBranches {
+		_, err := client.PutBranch(wharfapi.Branch{
+			Name:      branch.Name,
+			ProjectID: projectInDb.ProjectID,
+			Default:   branch.Ref == repositories.Value[0].DefaultBranch,
+			TokenID:   i.TokenID,
+		})
+		if err != nil {
+			fmt.Println("Unable to post branch: ", err)
+			ginutil.WriteAPIClientWriteError(c, err, fmt.Sprintf("Unable to import branch '%s'", branch.Name))
```

Comment:
nit: There's a handy `%q` format that I personally enjoy which quotes strings for you. It will escape quotes in the string, but I think it's fine, especially as some of these problems might get reported back to us developers where escaped text can be useful.

```suggestion
			ginutil.WriteAPIClientWriteError(c, err, fmt.Sprintf("Unable to import branch %q", branch.Name))
```